pkg/services/authn/clients/oauth.go

```diff
@@ -166,6 +168,15 @@ func (c *OAuth) Authenticate(ctx context.Context, r *authn.Request) (*authn.Iden
 	}
 	token.TokenType = "Bearer"
 
+	if oauthCfg.UseRefreshToken && token.RefreshToken == "" {
```

Comment:
I think the same check should be added to https://github.com/grafana/grafana/blob/main/pkg/services/oauthtoken/oauth_token.go#L446 as well to verify that the new refresh token is returned after a refresh.

Reply:
I was debating whether to do that and ultimately decided against it after reading through how `TokenSource` works ([1])[https://github.com/golang/oauth2/blob/f28b0b5467dda26d56f1240381158f7c334654d1/oauth2.go#L248-L261] [[2]](https://github.com/golang/oauth2/blob/f28b0b5467dda26d56f1240381158f7c334654d1/oauth2.go#L270-L291). At a glance, it seemed as though if the token used to make the token exchange request has a refresh token, it would either:

1. Be set to the current refresh token; or
2. Be set to the new refresh token (if it's different).

I suppose the key here is that it might be possible for it to be different and **empty**. I'll go ahead and at least log something when it happens from there.